File: design/src/main/java/com/kimjisub/design/panel/MainPackPanel.java
package com.kimjisub.design.panel;

import android.content.Context;
import android.content.res.TypedArray;
import android.databinding.DataBindingUtil;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.RelativeLayout;

import com.kimjisub.design.R;
import com.kimjisub.design.databinding.PanelMainPackBinding;

public class MainPackPanel extends RelativeLayout {

	public PanelMainPackBinding b;

	public MainPackPanel(Context context) {
		super(context);
		init();
	}

	public MainPackPanel(Context context, AttributeSet attrs) {
		super(context, attrs);
		init();
		getAttrs(attrs);
	}

	public MainPackPanel(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
		init();
		getAttrs(attrs, defStyleAttr);
	}

	private void init() {
		b = DataBindingUtil.inflate(LayoutInflater.from(getContext()), R.layout.panel_main_pack, this, true);

		b.title.setSelected(true);
		b.subTitle.setSelected(true);
		b.path.setSelected(true);


		b.star.setOnClickListener(v -> {
			if (onEventListener != null)
				onEventListener.onStarClick(v);
		});
		b.bookmark.setOnClickListener(v -> {
			if (onEventListener != null)
				onEventListener.onBookmarkClick(v);
		});
		b.edit.setOnClickListener(v -> {
			if (onEventListener != null)
				onEventListener.onEditClick(v);
		});
		b.storage.setOnClickListener(v -> {
			if (onEventListener != null)
				onEventListener.onStorageClick(v);
		});
		b.youtube.setOnClickListener(v -> {
			if (onEventListener != null)
				onEventListener.onYoutubeClick(v);
		});
		b.website.setOnClickListener(v -> {
			if (onEventListener != null)
				onEventListener.onWebsiteClick(v);
		});
		b.func.setOnClickListener(v -> {
			if (onEventListener != null)
				onEventListener.onFuncClick(v);
		});
		b.delete.setOnClickListener(v -> {
			if (onEventListener != null)
				onEventListener.onDeleteClick(v);
		});
	}

	private void getAttrs(AttributeSet attrs) {
		TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.MainPackPanel);
		setTypeArray(typedArray);
	}

	private void getAttrs(AttributeSet attrs, int defStyle) {
		TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.MainPackPanel, defStyle, 0);
		setTypeArray(typedArray);
	}

	private void setTypeArray(TypedArray typedArray) {

		typedArray.recycle();
	}

	////////////////////////////////////////////////////////////////////////////////////////////////

	public void setStar(boolean star) {
		b.star.setImageResource(star ? R.drawable.ic_star_24dp : R.drawable.ic_star_border_24dp);
	}

	public void setBookmark(boolean bookmark) {
		b.star.setImageResource(bookmark ? R.drawable.ic_bookmark_24dp : R.drawable.ic_bookmark_border_24dp);
	}

	public void setStorage(boolean external) {
		b.star.setImageResource(external ? R.drawable.ic_public_24dp : R.drawable.ic_lock_24dp);
		b.star.setClickable(true);
	}

	public void setStorageMoving() {
		b.star.setImageResource(R.drawable.ic_copy_24dp);
		b.star.setClickable(false);
	}


	////////////////////////////////////////////////////////////////////////////////////////////////

	OnEventListener onEventListener;

	public interface OnEventListener {
		void onStarClick(View v);

		void onBookmarkClick(View v);

		void onEditClick(View v);

		void onStorageClick(View v);

		void onYoutubeClick(View v);

		void onWebsiteClick(View v);

		void onFuncClick(View v);

		void onDeleteClick(View v);
	}

	public void setOnEventListener(OnEventListener onEventListener) {
		this.onEventListener = onEventListener;
	}
}

File: design/src/main/java/com/kimjisub/design/panel/MainPackPanel.kt
package com.kimjisub.design.panel

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.RelativeLayout
import androidx.databinding.ObservableField
import com.kimjisub.design.R.styleable
import com.kimjisub.design.databinding.PanelMainPackBinding
import java.util.*

class MainPackPanel
@JvmOverloads constructor(
	context: Context,
	attrs: AttributeSet? = null,
	defStyleAttr: Int = 0
) : RelativeLayout(context, attrs, defStyleAttr) {
	private val b: PanelMainPackBinding =
		PanelMainPackBinding.inflate(LayoutInflater.from(context), this, true)
	val data = Data()

	class Data {
		// Static
		val bookmark: ObservableField<Boolean> = ObservableField()
		val title: ObservableField<String> = ObservableField()
		val subtitle: ObservableField<String> = ObservableField()
		val padSize: ObservableField<String> = ObservableField()
		val chainCount: ObservableField<String> = ObservableField()
		val soundCount: ObservableField<String> = ObservableField()
		val ledCount: ObservableField<String> = ObservableField()
		val fileSize: ObservableField<String> = ObservableField()
		val playCount: ObservableField<String> = ObservableField()
		val downloadedDate: ObservableField<Date> = ObservableField()
		val lastPlayed: ObservableField<Date> = ObservableField()
		val websiteExist: ObservableField<Boolean> = ObservableField()
		val path: ObservableField<String> = ObservableField()
	}

	init {
		b.data = data

		attrs?.let {
			val typedArray =
				context.obtainStyledAttributes(it, styleable.StorePackPanel, defStyleAttr, 0)

			b.title.isSelected = true
			b.subtitle.isSelected = true
			b.path.isSelected = true
			b.bookmark.setOnClickListener { v: View ->
				onEventListener?.onBookmarkClick(v)
			}
			b.btnEdit.setOnClickListener { v: View ->
				onEventListener?.onEditClick(v)
			}
			b.btnYoutube.setOnClickListener { v: View ->
				onEventListener?.onYoutubeClick(v)
			}
			b.btnWebsite.setOnClickListener { v: View ->
				onEventListener?.onWebsiteClick(v)
			}
			b.btnFunc.setOnClickListener { v: View ->
				onEventListener?.onFuncClick(v)
			}
			b.btnDelete.setOnClickListener { v: View ->
				onEventListener?.onDeleteClick(v)
			}
			b.path.setOnClickListener { v: View ->
				onEventListener?.onPathClick(v)
			}

			typedArray.recycle()
		}
	}

	////////////////////////////////////////////////////////////////////////////////////////////////


	var onEventListener: OnEventListener? = null

	interface OnEventListener {
		fun onBookmarkClick(v: View)
		fun onEditClick(v: View)
		fun onYoutubeClick(v: View)
		fun onWebsiteClick(v: View)
		fun onFuncClick(v: View)
		fun onDeleteClick(v: View)
		fun onPathClick(v: View)
	}
}

